<a href="https://colab.research.google.com/github/mrsferret/Code-Division/blob/main/Worksheets/15_1_Encoding_and_Dummy_Coding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Encoding Data  
When we encode data, we remodel a numeric column in a dataframe to be on a standard scale (0 or 1).   
For example if we had a column of BMI scores, we could encode that column so that all scores greater than or equal to 25 were recoded to the value 1 (bad) and all scores less than 25 were recoded to 0 (good).  

For example:  

` def encode_bmi(df):`       
> `if df['bmi'] >= 25:`  
> >  `return 1`   

>`else`:  
> >`return 0`  

`df["bmi"] = df.apply(encode_bmi, axis=1)`    

### Challenge 1 - prepare dataset for encoding 
---
1. Read Covid vaccination data from the `by_country` sheet in the Excel file at this link : https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true
2. Find out which columns have missing values
3. Remove all rows with missing data in the total_vaccination column  
4. Remove all rows with missing data in the daily_vaccinations_per_million 
4. find the median daily_vaccinations_per_million, storing this in a variable for use later     


**Test output**:  
1. dataframe is saved in a variable
2. 
```
RangeIndex: 14994 entries, 0 to 14993
Data columns (total 15 columns):
    Column                               Non-Null Count  Dtype         
                                
 0   country                              14994 non-null  object        
 1   iso_code                             14994 non-null  object        
 2   date                                 14994 non-null  datetime64[ns]
 3   total_vaccinations                   9011 non-null   float64       
 4   people_vaccinated                    8370 non-null   float64       
 5   people_fully_vaccinated              6158 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   14796 non-null  float64       
 8   total_vaccinations_per_hundred       9011 non-null   float64       
 9   people_vaccinated_per_hundred        8370 non-null   float64       
 10  people_fully_vaccinated_per_hundred  6158 non-null   float64       
 11  daily_vaccinations_per_million       14796 non-null  float64       
 12  vaccines                             14994 non-null  object        
 13  source_name                          14994 non-null  object        
 14  source_website                       14994 non-null  object        
dtypes: datetime64[ns](1), float64(9), object(5)
memory usage: 1.7+ MB
```
3. 9011 rows × 15 columns  
4. 8815 rows * 15 columns 
5. 6.65 



In [37]:
import pandas as pd
url = "https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true"

# Task 2
def inspect_data(df):
  display("Describe: ", df.describe())
  display("Head: ", df.head())
  display("Info: ", df.info())

  # check the data for NA/NaN/missing values across the dataframe
  display("Null Values: ", df.isnull().values.any())

  cleaned_df = df

  return cleaned_df


# Task 3 & 4
def clean_data(df, col_name):

  # ================================================================
  # remove rows where col total_vaccinations, daily_vaccinations_per_million is null
  # ================================================================
  #
  clean_df = df[~df[col_name].isnull()]
  
  # clean_df = df.dropna(subset=['daily_vaccinations_per_million', 'total_vaccinations'], how = 'any')
  #clean_df = clean_df[~clean_df["daily_vaccinations_per_million"].isnull()]

  return clean_df
#=======================================================================
# Task 5
# find the median daily_vaccinations_per_million
#=======================================================================
def get_median(df):
  dly_vax_per_million_median = df['daily_vaccinations_per_million'].median()
  
  print("\ndly-vax_per_million_median = ", dly_vax_per_million_median)

  return dly_vax_per_million_median

#=======================================================================
# program starts here
#=======================================================================
# Task 1 - excel file into df
#=======================================================================
covid_vax_df = pd.read_excel(url)

print ("\ndly_vax_per_million_median 1: ", covid_vax_df.median())
min_total_vax = covid_vax_df[["country", "total_vaccinations"]].groupby("country").min()
print(min_total_vax.to_string())

#=======================================================================
# Task 2 - inspect data
#=======================================================================
inspect_df = inspect_data(covid_vax_df)

#=======================================================================
# Task 3 - remove rows where total_vaccinations is empty
#=======================================================================
col_name = 'total_vaccinations'
covid_vax_df = clean_data(inspect_df, col_name)

display("Info after total_vaccinations drop na: ", covid_vax_df.info())

#=======================================================================
# Task 4 - remove rows where daily_vaccinations_per_million is empty
#=======================================================================
col_name = 'daily_vaccinations_per_million'
covid_vax_df = clean_data(covid_vax_df, col_name)

display("Info after daily_vaccinations_per_million drop na: ", covid_vax_df.info())

#=======================================================================
# Task 5
# find the median daily_vaccinations_per_million
#=======================================================================
median = get_median(covid_vax_df)




dly_vax_per_million_median 1:  total_vaccinations                     420940.000
people_vaccinated                      325503.500
people_fully_vaccinated                172917.500
daily_vaccinations_raw                  15583.000
daily_vaccinations                       6178.500
total_vaccinations_per_hundred              6.300
people_vaccinated_per_hundred               4.825
people_fully_vaccinated_per_hundred         2.350
daily_vaccinations_per_million           1475.000
dtype: float64
                                  total_vaccinations
country                                             
Afghanistan                                      0.0
Albania                                          0.0
Algeria                                          0.0
Andorra                                        576.0
Angola                                           0.0
Anguilla                                         0.0
Antigua and Barbuda                              0.0
Argentina                 

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:49: FutureWarning: DataFrame.mean and DataFrame.median with numeric_only=None will include datetime64 and datetime64tz columns in a future version.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:49: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.


'Describe: '

total_vaccinations  people_vaccinated  people_fully_vaccinated  \
count        9.011000e+03       8.370000e+03             6.158000e+03   
mean         4.964584e+06       3.164030e+06             1.580482e+06   
std          2.033452e+07       1.180919e+07             6.840161e+06   
min          0.000000e+00       0.000000e+00             1.000000e+00   
25%          6.000000e+04       5.154925e+04             2.350025e+04   
50%          4.209400e+05       3.255035e+05             1.729175e+05   
75%          1.932222e+06       1.338161e+06             6.887102e+05   
max          2.753380e+08       1.470470e+08             1.047747e+08   

       daily_vaccinations_raw  daily_vaccinations  \
count            7.575000e+03        1.479600e+04   
mean             1.347569e+05        7.769163e+04   
std              5.211912e+05        3.519679e+05   
min              0.000000e+00        0.000000e+00   
25%              2.888500e+03        8.967500e+02   
50%              1.558300e+04        6.178500e+03   
75%              6.249850e+04        2.835000e+04   
max              1.160100e+07        7.205286e+06   

       total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
count                     9011.000000                    8370.000000   
mean                        15.456079                      11.228710   
std                         23.238675                      15.269686   
min                          0.000000                       0.000000   
25%                          1.330000                       1.222500   
50%                          6.300000                       4.825000   
75%                         19.570000                      14.667500   
max                        211.080000                     111.320000   

       people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  
count                           6158.00000                    14796.000000  
mean                               5.66235                     2972.163625  
std                               10.06184                     4720.846894  
min                                0.00000                        0.000000  
25%                                0.63000                      347.000000  
50%                                2.35000                     1475.000000  
75%                                6.30000                     3908.250000  
max                               99.76000                   118759.000000

'Head: '

country iso_code       date  total_vaccinations  people_vaccinated  \
0  Afghanistan      AFG 2021-02-22                 0.0                0.0   
1  Afghanistan      AFG 2021-02-23                 NaN                NaN   
2  Afghanistan      AFG 2021-02-24                 NaN                NaN   
3  Afghanistan      AFG 2021-02-25                 NaN                NaN   
4  Afghanistan      AFG 2021-02-26                 NaN                NaN   

   people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
0                      NaN                     NaN                 NaN   
1                      NaN                     NaN              1367.0   
2                      NaN                     NaN              1367.0   
3                      NaN                     NaN              1367.0   
4                      NaN                     NaN              1367.0   

   total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                             0.0                            0.0   
1                             NaN                            NaN   
2                             NaN                            NaN   
3                             NaN                            NaN   
4                             NaN                            NaN   

   people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
0                                  NaN                             NaN   
1                                  NaN                            35.0   
2                                  NaN                            35.0   
3                                  NaN                            35.0   
4                                  NaN                            35.0   

             vaccines                source_name  \
0  Oxford/AstraZeneca  Government of Afghanistan   
1  Oxford/AstraZeneca  Government of Afghanistan   
2  Oxford/AstraZeneca  Government of Afghanistan   
3  Oxford/AstraZeneca  Government of Afghanistan   
4  Oxford/AstraZeneca  Government of Afghanistan   

                                      source_website  
0  https://reliefweb.int/report/afghanistan/afgha...  
1  https://reliefweb.int/report/afghanistan/afgha...  
2  https://reliefweb.int/report/afghanistan/afgha...  
3  https://reliefweb.int/report/afghanistan/afgha...  
4  https://reliefweb.int/report/afghanistan/afgha...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14994 entries, 0 to 14993
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   country                              14994 non-null  object        
 1   iso_code                             14994 non-null  object        
 2   date                                 14994 non-null  datetime64[ns]
 3   total_vaccinations                   9011 non-null   float64       
 4   people_vaccinated                    8370 non-null   float64       
 5   people_fully_vaccinated              6158 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   14796 non-null  float64       
 8   total_vaccinations_per_hundred       9011 non-null   float64       
 9   people_vaccinated_per_hundred        8370 non-null   float64       
 10  people_ful

'Info: '

None

'Null Values: '

True

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9011 entries, 0 to 14993
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   country                              9011 non-null   object        
 1   iso_code                             9011 non-null   object        
 2   date                                 9011 non-null   datetime64[ns]
 3   total_vaccinations                   9011 non-null   float64       
 4   people_vaccinated                    8294 non-null   float64       
 5   people_fully_vaccinated              6157 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   8816 non-null   float64       
 8   total_vaccinations_per_hundred       9011 non-null   float64       
 9   people_vaccinated_per_hundred        8294 non-null   float64       
 10  people_full

'Info after total_vaccinations drop na: '

None

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8816 entries, 6 to 14993
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   country                              8816 non-null   object        
 1   iso_code                             8816 non-null   object        
 2   date                                 8816 non-null   datetime64[ns]
 3   total_vaccinations                   8816 non-null   float64       
 4   people_vaccinated                    8126 non-null   float64       
 5   people_fully_vaccinated              6143 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   8816 non-null   float64       
 8   total_vaccinations_per_hundred       8816 non-null   float64       
 9   people_vaccinated_per_hundred        8126 non-null   float64       
 10  people_full

'Info after daily_vaccinations_per_million drop na: '

None


dly-vax_per_million_median =  1915.5


**Task 2**

From task 2's output, can see that there are 14994 rows and 15 columns in dataframe.  The following 9 columns have missing data:                       

*   total_vaccinations
*   people_vaccinated
*   people_fully_vaccinated
*   daily_vaccinations_raw 
*   daily_vaccinations
*   total_vaccinations_per_hundred
*   people_vaccinated_per_hundred
*   people_fully_vaccinated_per_hundred 
*   daily_vaccinations_per_million 


### Challenge 2 - encode daily vaccinations 
---

Write a function to encode daily vaccinations per million, where values greater than or equal to median = 1 and values less than median = 0 

**Test output**: 

using describe()
```
count    8816.000000
mean        0.991493
std         0.091847
min         0.000000
25%         1.000000
50%         1.000000
75%         1.000000
max         1.000000
Name: daily_vaccinations_per_million, dtype: float64
```

In [38]:
# Task
def encode_daily(df, **kwds):
  median = kwds['median']
  key = kwds['key']

  if df[key] >= median:
    return 1
  else:
    return 0

#=============================================================================
# Start
#=======================================================================
# find the median daily_vaccinations_per_million
#=======================================================================
median = get_median(covid_vax_df)
covid_vax_df["encoded"] = covid_vax_df.apply(encode_daily, axis = 1, key="daily_vaccinations_per_million", median=median)

covid_vax_df['encoded'].describe()




dly-vax_per_million_median =  1915.5


count    8816.000000
mean        0.500000
std         0.500028
min         0.000000
25%         0.000000
50%         0.500000
75%         1.000000
max         1.000000
Name: encoded, dtype: float64

### Challenge 3 - Encoding total vaccinations   
---
The United Kingdom has been praised for its fast vaccine rollout. 
1. Find the minimum total vaccinations for the United Kingdom 
2. Write a function to encode total_vaccinations column so that all values less than the UK's min are 0 and all values greater than or equal to the UK's min are coded as 1 
3. Display the unique countries for which total vaccinated is at the same rate or more than the UK

**Test output**:

1. 1402432.0
2. `df['total_vaccinations'].describe()` should output:
```
count    9011.00000
mean        0.29808
std         0.45744
min         0.00000
25%         0.00000
50%         0.00000
75%         1.00000
max         1.00000
Name: total_vaccinations, dtype: float64
```
3. 
```
array(['Argentina', 'Australia', 'Austria', 'Azerbaijan', 'Bangladesh',
       'Belgium', 'Brazil', 'Cambodia', 'Canada', 'Chile', 'China',
       'Colombia', 'Czechia', 'Denmark', 'Dominican Republic', 'England',
       'Finland', 'France', 'Germany', 'Greece', 'Hong Kong', 'Hungary',
       'India', 'Indonesia', 'Ireland', 'Israel', 'Italy', 'Japan',
       'Kazakhstan', 'Malaysia', 'Mexico', 'Morocco', 'Nepal',
       'Netherlands', 'Norway', 'Pakistan', 'Peru', 'Philippines',
       'Poland', 'Portugal', 'Qatar', 'Romania', 'Russia', 'Saudi Arabia',
       'Scotland', 'Serbia', 'Singapore', 'Slovakia', 'South Korea',
       'Spain', 'Sweden', 'Switzerland', 'Thailand', 'Turkey',
       'United Arab Emirates', 'United Kingdom', 'United States',
       'Uruguay', 'Wales'], dtype=object)
```




In [47]:
# Task 1

#=======================================================================
# Task 2 - encode total_vaccinations col:
# all values < UK min are 0 and otherwise coded as 1
#=======================================================================
def encode_total(df, **kwds):
  total = kwds['minimum']
  key = kwds['key']

  if df[key] >= total:
    return 1
  else:
    return 0

#=======================================================================
# Task 3 - Display the unique countries for which total vaccinated is 
# at the same rate or more than the UK
#=======================================================================
def display_unique_countries(df):

  df.encoded_2==1

  #??????????????????????????????????????????????????

  return


#=============================================================================
# Start
#=======================================================================
# Task 1 - find the min UK total_vaccinations 
#=======================================================================
# selecting rows based on condition

minimum = covid_vax_df[covid_vax_df['country'] == 'United Kingdom'].total_vaccinations.min()

print("\nMin UK total Vax: ", minimum)

#=======================================================================
# Task 2 - encode total_vaccinations col:
# all values < UK min are 0 and otherwise coded as 1
#=======================================================================
covid_vax_df["encoded_2"] = covid_vax_df.apply(encode_total, axis = 1, key="total_vaccinations", minimum=minimum)

covid_vax_df['encoded_2'].describe()

#=======================================================================
# Task 3 - Display the unique countries for which total vaccinated is 
# at the same rate or more than the UK
#=======================================================================
display_unique_countries(covid_vax_df )







Min UK total Vax:  2677971.0


### Challenge 4 - create new series of total vaccinations for each manufacturer
---

To create a new column in your dataframe:

`df['new_column'] = ...`

For example:

* to duplicate an existing column
  * `df['new_column'] = df['old_column']`
* to add two columns together 
  * `df['new_column'] = df['column1'] + df['column2']`
* to make a percentages column 
  * `df['new_column'] = (df['column1']/df['column1].sum()) * 100`

  
1. read data from 'by_manufacturer' sheet from Covid data 
2. find the sum of total vaccinations for each manufacturer
3. create a new column that has the total vaccinations as a percentage of the overall sum of total vaccinations 
4. find the median percentage 
5. create a new column called 'encoded_percentages' which duplicates the percentages column
6. encode the encoded_percentages column so that any values greater than or equal to the median percentage = 1 and any lesser than = 0 


**Test output**:

1.
2. 
```
vaccine
Johnson&Johnson        264839828
Moderna               5548036383
Oxford/AstraZeneca     539433203
Pfizer/BioNTech       8690461304
Sinovac                604660293
Name: total_vaccinations, dtype: int64
```
3. 
```
	location	date	vaccine	total_vaccinations	percentages
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055
...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519
3294	United States	2021-05-02	Moderna	106780082	0.682413
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423
3296 rows × 5 columns
```
4. 0.0011110194374896931
5. 
```
location	date	vaccine	total_vaccinations	percentage_of_total	encoded_percentages
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003	0.000003
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033	0.000033
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053	0.000053
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055	0.000055
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055	0.000055
...	...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095	0.677095
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504	0.824504
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519	0.053519
3294	United States	2021-05-02	Moderna	106780082	0.682413	0.682413
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423	0.832423
3296 rows × 6 columns
```
6. 
```
	location	date	vaccine	total_vaccinations	percentages	encode	encoded
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003	0.000003	0
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033	0.000033	0
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053	0.000053	0
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055	0.000055	0
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055	0.000055	0
...	...	...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095	0.677095	1
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504	0.824504	1
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519	0.053519	1
3294	United States	2021-05-02	Moderna	106780082	0.682413	0.682413	1
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423	0.832423	1
3296 rows × 7 columns
```



In [66]:

def calc_sum_per_mnfctr(df):
  sum_total_vax = df[["vaccine", "total_vaccinations"]].groupby("vaccine").sum()
  print(sum_total_vax.to_string())

  return sum_total_vax

def create_new_col(df):

  df['percent_total_sum_vax'] = (df['total_vaccinations']/df['total_vaccinations'].sum()) * 100
  #df['new_column'] = (df['column1']/df['column1].sum()) * 100
  print("\ndf: ", df.columns.values)
  #list(df.columns.values)
  return df
#=======================================================================
# Task 4
# find the median percent_total_sum_vax
#=======================================================================
def calc_median(df):

  median_percent = df['percent_total_sum_vax'].median()
  
  print("\median_percent = ", median_percent)

  return median_percent

#=======================================================================
# Task 5 - encode percent_total_sum_vax col:
# all values >= median percentage = 1 and otherwise coded as 0
#=======================================================================
def encode_total(df, **kwds):
  total = kwds['median']

  print("\ntotal: ", total)

  key = kwds['key']

  print("\nkey: ", key)

  test = df[key]
  print(test)

  if df[key] >= total:
    return 1
  else:
    return 0

#def encode_total(df, median):
#  
#  if df['percent_total_sum_vax'] >= median:
#    return 1
#  else:
#    return 0

#=======================================================================
# program starts here
#=======================================================================
# Task 1 - excel file into df
#=======================================================================

url = "https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true"
covid_vax_by_mnfctr_df = pd.read_excel(url,sheet_name="by_manufacturer")

display (covid_vax_by_mnfctr_df.columns)

print("\ncovid_vax_by_mnfctr_df describe: ", covid_vax_by_mnfctr_df.describe())

#=======================================================================
# Task 2 - find the sum of total vaccinations for each manufacturer
#=======================================================================
sum_total_vax = calc_sum_per_mnfctr(covid_vax_by_mnfctr_df)

#=======================================================================
# Task 3 - create new col with total vaccinations as a % overall sum of total vaccinations
#=======================================================================

covid_vax_by_mnfctr_new_df = create_new_col(covid_vax_by_mnfctr_df)

#=======================================================================
# Task 4 - find median percentage of new col percent_total_sum_vax
#=======================================================================

median_percent = calc_median(covid_vax_by_mnfctr_new_df)

print("\nmedian percent: ", median_percent)

#=======================================================================
# Task 5 - encode percent_total_sum_vax col:
#=======================================================================
covid_vax_by_mnfctr_new_df["encoded_percentages"] = covid_vax_df.apply(encode_total, axis = 1, key="percent_total_sum_vax", median=median_percent)

#covid_vax_by_mnfctr_new_df["encoded_percentages"] = covid_vax_by_mnfctr_new_df.apply(encode_total(covid_vax_by_mnfctr_new_df, median), axis = 1)

covid_vax_by_mnfctr_new_df['encoded_percentages'].describe()



Index(['location', 'date', 'vaccine', 'total_vaccinations'], dtype='object')


covid_vax_by_mnfctr_df describe:         total_vaccinations
count        3.296000e+03
mean         4.747400e+06
std          1.578003e+07
min          0.000000e+00
25%          1.571925e+04
50%          1.738460e+05
75%          1.790371e+06
max          1.302528e+08
                    total_vaccinations
vaccine                               
Johnson&Johnson              264839828
Moderna                     5548036383
Oxford/AstraZeneca           539433203
Pfizer/BioNTech             8690461304
Sinovac                      604660293

df:  ['location' 'date' 'vaccine' 'total_vaccinations' 'percent_total_sum_vax']
\median_percent =  0.0011110194374896931

median percent:  0.0011110194374896931

total:  0.0011110194374896931

key:  percent_total_sum_vax


KeyError: ignored

### Exercise 8 - encode daily vaccinations 

1. find the median daily vaccinations per 1 million 
2. write a function to encode daily vaccinations per 1 million, where values greater than or equal to median = 1 and values less than median = 0 

Output: 

1. 1915.5
2. 
```
0        0
6        0
22       0
44       0
59       0
        ..
14989    0
14990    0
14991    0
14992    0
14993    0
Name: daily_vaccinations_per_million, Length: 9011, dtype: int64
```

In [65]:
# Task 2 
#encode daily vaccinations per 1 million, where values greater than or equal to 
# median = 1 and values less than median = 0
def encode_daily_vax_per_million(df, **kwds):
  median = kwds['median']
  key = kwds['key']

  if df[key] >= median:
    return 1
  else:
    return 0

#=======================================================================
# Task 1
# find the median daily_vaccinations_per_million
#=======================================================================
def calc_median(df):

  #median_percent = df['daily_vaccinations_per_million'].median()
  
  print("\median_percent = ", median_percent)

  return median_percent

#=============================================================================
# Start
#=======================================================================
# find the median daily_vaccinations_per_million
#=======================================================================
median = calc_median(covid_vax_by_mnfctr_new_df)

display(covid_vax_by_mnfctr_new_df.columns)

#covid_vax_by_mnfctr_new_df["encoded"] = covid_vax_by_mnfctr_new_df.apply(encode_daily, axis = 1, key="daily_vaccinations_per_million", median=median)

#covid_vax_by_mnfctr_new_df['encoded'].describe()


\median_percent =  0.0011110194374896931


Index(['location', 'date', 'vaccine', 'total_vaccinations',
       'percent_total_sum_vax'],
      dtype='object')

### Exercise 9 - Encoding vaccinations per hundred  
---
The United Kingdom has been praised for its fast vaccine rollout. 
1. find the minimum total vaccinations for the United Kingdom 
2. save this value in a variable rounded down to an integer
3. write a function to encode total_vaccinations column so that all values less than the UK's min are 0 and all values greater than or equal to the UK's min are coded as 1 
4. display the countries which total vaccinated is at the same rate or more than the UK

Output:

1. 1402432.0
2. 1402432
3. `df['people_vaccinated_per_hundred']` should output:
```
0        0
6        0
22       0
44       0
59       0
        ..
14989    0
14990    0
14991    0
14992    0
14993    0
Name: total_vaccinations, Length: 9011, dtype: int64
```
4. 
```
array(['Argentina', 'Australia', 'Austria', 'Azerbaijan', 'Bangladesh',
       'Belgium', 'Brazil', 'Cambodia', 'Canada', 'Chile', 'China',
       'Colombia', 'Czechia', 'Denmark', 'Dominican Republic', 'England',
       'Finland', 'France', 'Germany', 'Greece', 'Hong Kong', 'Hungary',
       'India', 'Indonesia', 'Ireland', 'Israel', 'Italy', 'Japan',
       'Kazakhstan', 'Malaysia', 'Mexico', 'Morocco', 'Nepal',
       'Netherlands', 'Norway', 'Pakistan', 'Peru', 'Philippines',
       'Poland', 'Portugal', 'Qatar', 'Romania', 'Russia', 'Saudi Arabia',
       'Scotland', 'Serbia', 'Singapore', 'Slovakia', 'South Korea',
       'Spain', 'Sweden', 'Switzerland', 'Thailand', 'Turkey',
       'United Arab Emirates', 'United Kingdom', 'United States',
       'Uruguay', 'Wales'], dtype=object)
```




### Exercise 10 - create new series of total vaccinations percentages
---

To create a new column in your dataframe:

`df['new_column'] = ...`

For example:

* to duplicate an existing column
  * `df['new_column'] = df['old_column']`
* to add two columns together 
  * `df['new_column'] = df['column1'] + df['column2']`
* to make a percentages column 
  * `df['new_column'] = (df['column1']/df['column1].sum()) * 100`  
  


1. read data from 'by_manufacturer' sheet from Covid data 
2. find the sum of total vaccinations for each manufacturer
3. create a new column that has the total vaccinations as a percentage of the overall sum of total vaccinations 
4. find the median percentage 
5. create a new column called 'encoded_percentages' which duplicates the percentages column
6. encode the encoded_percentages column so that any values greater than or equal to the median percentage = 1 and any lesser than = 0 


Output:

1.
2. 
```
vaccine
Johnson&Johnson        264839828
Moderna               5548036383
Oxford/AstraZeneca     539433203
Pfizer/BioNTech       8690461304
Sinovac                604660293
Name: total_vaccinations, dtype: int64
```
3. 
```
	location	date	vaccine	total_vaccinations	percentages
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055
...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519
3294	United States	2021-05-02	Moderna	106780082	0.682413
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423
3296 rows × 5 columns
```
4. 0.0011110194374896931
5. 
6. 
```
	location	date	vaccine	total_vaccinations	percentages	encode	encoded
0	Chile	2020-12-24	Pfizer/BioNTech	420	0.000003	0.000003	0
1	Chile	2020-12-25	Pfizer/BioNTech	5198	0.000033	0.000033	0
2	Chile	2020-12-26	Pfizer/BioNTech	8338	0.000053	0.000053	0
3	Chile	2020-12-27	Pfizer/BioNTech	8649	0.000055	0.000055	0
4	Chile	2020-12-28	Pfizer/BioNTech	8649	0.000055	0.000055	0
...	...	...	...	...	...	...	...
3291	United States	2021-05-01	Moderna	105947940	0.677095	0.677095	1
3292	United States	2021-05-01	Pfizer/BioNTech	129013657	0.824504	0.824504	1
3293	United States	2021-05-02	Johnson&Johnson	8374395	0.053519	0.053519	1
3294	United States	2021-05-02	Moderna	106780082	0.682413	0.682413	1
3295	United States	2021-05-02	Pfizer/BioNTech	130252779	0.832423	0.832423	1
3296 rows × 7 columns
```



# Reflection
----

## What skills have you demonstrated in completing this notebook?

Your answer:

## What caused you the most difficulty?

Your answer: